From 45c00927b72319f00855c7feaf4dcf109b3e4414 Mon Sep 17 00:00:00 2001
From: Markus Volk <f_l_k@t-online.de>
Date: Fri, 8 Apr 2022 08:58:46 +0000
Subject: [PATCH] Use POSIX basename() in BCache_Info.cc (!99)

Musl libc [1][2] doesn't implement the GNU variant of basename() [3][4],
obtained via #include <string.h>.  Therefore GParted fails to build on
such distributions:

   fdebug-prefix-map=TOPDIR/build/tmp/work/cortexa57-yoe-linux-musl/gparted/1.4.0-r0/recipe-sysroot-native=-fvisibility-inlines-hidden  -c -o ../../gparted-1.4.0/src/BCache_Info.cc:52:33:

    error: use of undeclared identifier 'basename'; did you mean 'g_basename'?
            return "/dev/" + Glib::ustring(basename(buf));
                                           ^~~~~~~~
                                           g_basename

Fix by using the POSIX implementation of basename() [5] instead,
obtained via #include <libgen.h>, which musl libc does implement [6].
Note that the POSIX implementation of basename() is allowed to modify
the string passed to it.  This is okay because
BCache_Info::get_bcache_device() is using a modifiable local character
buffer.

[1] musl libc
    https://musl.libc.org/
[2] Projects using musl
    https://wiki.musl-libc.org/projects-using-musl.html
[3] The GNU C Library, 5.10 Finding Tokens in a String
    https://www.gnu.org/software/libc/manual/html_node/Finding-Tokens-in-a-String.html
[4] basename(3) - Linux manual page
    https://man7.org/linux/man-pages/man3/basename.3.html
[5] POSIX basename()
    https://pubs.opengroup.org/onlinepubs/009695399/functions/basename.html
[6] musl source, basename.c
    http://git.musl-libc.org/cgit/musl/tree/src/misc/basename.c

Closes !99 - Fix undeclared identifier 'basename' build failure with
             musl libc
---
 src/BCache_Info.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/BCache_Info.cc b/src/BCache_Info.cc
index 129a5178..aecc9ef1 100644
--- a/src/BCache_Info.cc
+++ b/src/BCache_Info.cc
@@ -17,7 +17,7 @@
 
 #include "BCache_Info.h"
 
-#include <string.h>  // GNU version of basename()
+#include <libgen.h>  // POSIX version of basename()
 #include <unistd.h>
 #include <glibmm/ustring.h>
 #include <glibmm/fileutils.h>
-- 
GitLab

